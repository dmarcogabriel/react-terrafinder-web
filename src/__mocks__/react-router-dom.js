const originalModule = jest.requireActual('react-router-dom');

module.exports = {
  ...originalModule,
  useHistory: () => ({
    history: { push: jest.fn() },
  }),
};
