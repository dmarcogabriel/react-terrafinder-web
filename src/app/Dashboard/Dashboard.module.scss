@import 'styles';

.container {
  display: flex;
}

.modal {
  color: $gray-dark;

  > h1 {
    font-size: 2rem;
  }

  > p {
    line-height: 1.5rem;
    margin: 1rem 0;
  }

  > img {
    width: 100%;
    max-height: 150px;
  }

  .buttons {
    display: flex;
    flex-direction: column;
  }

  > input {
    display: none;
  }
}
