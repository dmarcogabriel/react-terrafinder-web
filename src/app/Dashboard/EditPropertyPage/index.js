import { EditPropertyPage } from './EditPropertyPage';

export default EditPropertyPage;
