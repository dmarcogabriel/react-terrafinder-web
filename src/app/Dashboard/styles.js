import styled from 'styled-components';
import CommonButton from 'common/components/atm/Button';

export const Button = styled(CommonButton)`
  margin: 1rem 0;
`;
