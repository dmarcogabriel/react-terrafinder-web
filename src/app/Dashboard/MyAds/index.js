import React, { useEffect, useState } from 'react';
import Property from 'common/components/Property';
import { MdAdd } from 'react-icons/md';
import { useUser } from 'hooks/useUser';
import api from 'services/api';
import { useHistory } from 'react-router-dom';
import classes from './MyAds.module.scss';
import { CreateAdButton } from './styles';

export default function MyAds() {
  const { currentUser } = useUser();
  const history = useHistory();
  const [myProperties, setMyProperties] = useState([]);

  const handleSelect = (id) => {
    history.push(`property/${id}`);
  };

  const loadMyProperties = async () => {
    try {
      console.log('User loggado', currentUser);
      const { data: res } = await api.get(`properties/user/${currentUser._id}`);
      setMyProperties(res.data.properties);
    } catch (error) {
      setMyProperties([]);
    }
  };

  const newPropertyAd = () => history.push('/create/property?step=1');

  useEffect(() => {
    loadMyProperties();
  }, []);

  return (
    <div className={classes.myAds}>
      <h1>Meus anúncios</h1>

      <CreateAdButton dataTestId="createProperty" onClick={newPropertyAd}>
        <MdAdd size={22} />

        <p>Criar Anúncio</p>
      </CreateAdButton>

      <div data-testid="propertyList" className={classes.adsList}>
        {myProperties.map((property, i) => (
          <Property
            dataTestId={`property-${i}`}
            key={property._id}
            index={i}
            property={property}
            onSelect={handleSelect}
          />
        ))}
      </div>
    </div>
  );
}
