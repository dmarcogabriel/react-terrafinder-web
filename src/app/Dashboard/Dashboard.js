import React, { useEffect, useState, useRef } from 'react';
import Header from 'common/components/Header';
import Footer from 'common/components/Footer';
import Modal from 'common/components/Modal';
import { Route, Switch } from 'react-router-dom';
import { useUser } from 'hooks/useUser';
import api from 'services/api';
import { loadBase64Image } from 'utils/fileHelper';
import { useTheme } from 'hooks/useTheme';
import classes from './Dashboard.module.scss';
import MyAds from './MyAds';
import { Button } from './styles';

export const Dashboard = () => {
  const { currentUser, updateUser } = useUser();
  const [showModal, setShowModal] = useState(false);
  const [avatar, setAvatar] = useState();
  const inputRef = useRef();
  const { color } = useTheme();

  const openGalery = () => {
    inputRef.current.click();
  };

  const handleCloseModal = () => setShowModal(false);

  const handleFileInputChange = async (e) => {
    const [file] = e.target.files;

    file.data = await loadBase64Image(file);

    setAvatar(file);
  };

  const uploadPhoto = async () => {
    try {
      const data = new FormData();

      data.append('avatar', avatar);

      const res = await api.put(
        `users/upload-photos/${currentUser._id}`,
        data,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      );

      updateUser({ avatar: res.data.avatar });
      setShowModal(false);
    } catch (error) {
      console.log('erro', error.response);
    }
  };

  useEffect(() => {
    console.log('Current User: ', currentUser);
    if (!currentUser.avatar) {
      setShowModal(true);
    }
  }, []);

  return (
    <>
      <div className={classes.dashboard}>
        <Header />

        <Switch>
          <Route path="/">
            <MyAds />
          </Route>
        </Switch>

        <Footer />
      </div>

      <Modal show={showModal}>
        <div className={classes.modal}>
          <h1>{`Bem vindo, ${currentUser.firstName}!`}</h1>

          {avatar ? (
            <p>Confirma essa foto?</p>
          ) : (
            <p>
              Seu perfil está quase completo! Adicione uma foto sua para chamar
              a atenção dos futuros compradores.
            </p>
          )}

          {avatar && <img src={avatar.data} alt="avatar" />}

          <div className={classes.buttons}>
            {avatar ? (
              <Button
                modifiers="custom"
                bg={color.blue.df}
                color={color.white.df}
                onClick={uploadPhoto}
              >
                Confirmar
              </Button>
            ) : (
              <Button
                modifiers="custom"
                bg={color.blue.df}
                color={color.white.df}
                dataTestId="uploadPhoto"
                onClick={openGalery}
              >
                Upload de foto
              </Button>
            )}

            <Button modifiers="secondary" onClick={handleCloseModal}>
              {avatar ? 'Cancelar' : 'Agora não'}
            </Button>
          </div>

          <input
            data-testid="fileInput"
            type="file"
            ref={inputRef}
            accept="image/png, image/jpeg"
            onChange={handleFileInputChange}
          />
        </div>
      </Modal>
    </>
  );
};
