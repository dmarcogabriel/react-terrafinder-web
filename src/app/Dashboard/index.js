import React, { useEffect, useState, useRef } from 'react';
import Header from 'common/components/Header';
import Footer from 'common/components/Footer';
import Button from 'common/components/Button';
import Modal from 'common/components/Modal';
import { Route, Switch } from 'react-router-dom';
import { useUser } from 'hooks/useUser';
import api from 'services/api';
import { loadBase64Image } from 'utils/fileHelper';
import classes from './Dashboard.module.scss';
import MyAds from './MyAds';

export default function Dashboard() {
  const { currentUser, updateUser } = useUser();
  const [showModal, setShowModal] = useState(false);
  const [avatar, setAvatar] = useState();
  const inputRef = useRef();

  const openGalery = () => {
    inputRef.current.click();
  };

  const handleCloseModal = () => setShowModal(false);

  const handleFileInputChange = async (e) => {
    const [file] = e.target.files;

    file.data = await loadBase64Image(file);

    setAvatar(file);
  };

  const uploadPhoto = async () => {
    try {
      const data = new FormData();

      data.append('avatar', avatar);

      const res = await api.put(
        `users/upload-photos/${currentUser._id}`,
        data,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
            'x-access-token': currentUser.token,
          },
        }
      );

      updateUser({ avatar: res.data.avatar });
      setShowModal(false);
    } catch (error) {
      console.log('erro', error.response);
    }
  };

  useEffect(() => {
    if (!currentUser.avatar) {
      setShowModal(true);
    }
  }, []);

  return (
    <>
      <div className={classes.dashboard}>
        <Header />

        <Switch>
          <Route path="/">
            <MyAds />
          </Route>
        </Switch>

        <Footer />
      </div>

      <Modal show={showModal}>
        <div className={classes.modal}>
          <h1>{`Bem vindo, ${currentUser.firstName}!`}</h1>

          {avatar ? (
            <p>Confirma essa foto?</p>
          ) : (
            <p>
              Seu perfil está quase completo! Adicione uma foto sua para chamar
              a atenção dos futuros compradores.
            </p>
          )}

          {avatar && <img src={avatar.data} alt="avatar" />}

          <div className={classes.buttons}>
            {avatar ? (
              <Button className={classes.button} onClick={uploadPhoto}>
                Confirmar
              </Button>
            ) : (
              <Button
                dataTestId="uploadPhoto"
                className={classes.button}
                onClick={openGalery}
              >
                Upload de foto
              </Button>
            )}

            <Button className={classes.cancelButton} onClick={handleCloseModal}>
              {avatar ? 'Cancelar' : 'Agora não'}
            </Button>
          </div>

          <input
            data-testid="fileInput"
            type="file"
            ref={inputRef}
            accept="image/png, image/jpeg"
            onChange={handleFileInputChange}
          />
        </div>
      </Modal>
    </>
  );
}
