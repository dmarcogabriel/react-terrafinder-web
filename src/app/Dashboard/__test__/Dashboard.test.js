import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import Dashboard from '..';

const Comp = () => (
  <Router>
    <Dashboard />
  </Router>
);

jest.mock('hooks/useUser', () => ({
  useUser: () => ({
    currentUser: {
      _id: 'testId',
      firstName: 'Tester',
      avatar: null,
    },
  }),
}));

describe('<Dashboard />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Comp />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should open galery', () => {
    const { getByTestId } = render(<Comp />);

    const fileInput = getByTestId('fileInput');
    const spy = jest.spyOn(fileInput, 'click');

    const uploadPhoto = getByTestId('uploadPhoto');
    fireEvent.click(uploadPhoto);

    expect(spy).toHaveBeenCalled();
  });

  // todo: test file upload

  it('matches snapshot', () => {
    const tree = renderer.create(<Comp />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
