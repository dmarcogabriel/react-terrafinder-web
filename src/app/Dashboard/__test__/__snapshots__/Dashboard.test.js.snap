// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> matches snapshot 1`] = `
.c0 {
  cursor: pointer;
  text-align: center;
  font-family: Roboto;
  box-shadow: 0 0.2rem 0.6rem 0 rgba(0,0,0,0.7);
  padding: 0.75rem;
  border-radius: 0.25rem;
  background: #62ac43;
  color: #fff;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 1rem 0;
}

.c1 > p {
  margin-left: 1rem;
}

@media (min-width:768px) {
  .c1 {
    position: absolute;
    right: 2rem;
    top: 9rem;
  }
}

<div>
  <div>
    <div>
      <a
        href="/home"
        onClick={[Function]}
      >
        <img
          alt="logo"
          src="\\"Logo.png\\""
        />
      </a>
      <button
        data-testid="menuButton"
        onClick={[Function]}
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </button>
      <a
        href="/home"
        onClick={[Function]}
      >
        <img
          alt="logo desktop"
          src="\\"Logo.svg\\""
        />
      </a>
      <p
        data-testid="greetings"
      >
        Olá, Tester
      </p>
      <button
        data-testid="logoutButton"
        onClick={[Function]}
        type="button"
      >
        Logout
      </button>
    </div>
  </div>
  <div>
    <h1>
      Meus anúncios
    </h1>
    <button
      className="c0 c1"
      data-testid="createProperty"
      onClick={[Function]}
      type="button"
    >
      <svg
        fill="currentColor"
        height={22}
        size={22}
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width={22}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <p>
        Criar Anúncio
      </p>
    </button>
    <div
      data-testid="propertyList"
    />
  </div>
  <footer>
    <hr />
    <div>
      <div>
        <p>
          Terrafinder
        </p>
      </div>
      <div>
        <p>
          Institucional
        </p>
        <ul>
          <li>
            <a
              href="/home"
              onClick={[Function]}
            >
              Página Inicial
            </a>
          </li>
          <li>
            <a
              href="/home/about"
              onClick={[Function]}
            >
              Quem somos
            </a>
          </li>
          <li>
            <a
              href="/home/privacy-policy"
              onClick={[Function]}
            >
              Política de privacidade
            </a>
          </li>
          <li>
            <a
              href="/home/support"
              onClick={[Function]}
            >
              Fale Conosco
            </a>
          </li>
        </ul>
      </div>
      <div>
        <p>
          Funcionalidades
        </p>
        <ul>
          <li>
            <a
              href="/login"
              onClick={[Function]}
            >
              Quero anunciar meu imóvel
            </a>
          </li>
          <li>
            <a
              href="/home/search-property"
              onClick={[Function]}
            >
              Quero comprar um imóvel
            </a>
          </li>
          <li>
            <a
              href="/home"
              onClick={[Function]}
            >
              Preços e planos
            </a>
          </li>
        </ul>
      </div>
      <div>
        <p>
          Fique por dentro das novidades!
        </p>
        <p>
          Se inscreva em nossa newsletter para receber os anúncios com os melhores preços!
        </p>
        <form>
          <input
            data-testid="emailInput"
            onChange={[Function]}
            placeholder="Endereço de email"
            type="email"
            value=""
          />
          <button
            data-testid="registerEmail"
            onClick={[Function]}
            type="button"
          >
            Cadastrar
          </button>
        </form>
      </div>
    </div>
    <hr />
    <p>
      Copyright C 2020 Terrafinder. Todos os direitos reservados.
    </p>
  </footer>
</div>
`;
